docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,332 @@
+```
+
+This would add a user supplied compression algorithm
+
+### Client Configuration
+
+This feature can be configured by a customer to adjust the minimum payload size threshold used to determine when a request should be compressed or be disabled completely. These settings are sourced from the following locations in priority order:
+
+1. Explicit client configuration
+2. JVM system properties
+3. Environment variables
+4. AWS profile
+
+*NOTE*: The client needs to be instantiated using `fromEnvironment()` for it to check all the available locations or else it
+will only look at the explicit client config.
+
+Will look in all 4 locations:
+```kotlin
+    val client = S3Client.fromEnvironment {
+        region = "us-west-2"
+    }
+```
+
+Will only look at explicit client config:
+```kotlin
+    val client = S3Client {
+        region = "us-west-2"
+        disableRequestCompression = true
+    }
+```
+
+#### Disable Request Compression 
+* Type: Boolean
+* Default: False
+
+
+| Description                  | Client configuration             | JVM system property                  | Environment variable                     | AWS profile key                      |
+|------------------------------|----------------------------------|--------------------------------------|------------------------------------------|--------------------------------------|
+| Enable/disable compression   | `disableRequestCompression`      | `aws.disableRequestCompression`      | `AWS_DISABLE_REQUEST_COMPRESSION`        | `disable_request_compression`        |
+
+#### Minimum Request Compression Size Bytes 
+* Type: Int
+* Default: 10,240 (10KB)
+* Range: 0 - 10,485,760 (0MB - 10MB) 
+* Notes: Setting will be ignored if payload is a stream
+
+| Description                  | Client configuration             | JVM system property                  | Environment variable                     | AWS profile key                      |
+|------------------------------|----------------------------------|--------------------------------------|------------------------------------------|--------------------------------------|
+| Min payload size to compress | `requestMinCompressionSizeBytes` | `aws.requestMinCompressionSizeBytes` | `AWS_REQUEST_MIN_COMPRESSION_SIZE_BYTES` | `request_min_compression_size_bytes` |
+
+This feature can also be configured to use user supplied compression algorithms. The user can add compression algorithms using the client configuration function `addCompressionAlgorithms`. A user supplied algorithm may override the built-in implementation, and the compression algorithm must implement the `CompressionAlgorithm` interface.
+```kotlin
+val client = S3Client.fromEnvironment {
+    region = "us-west-2"
+}
+client.addCompressionAlgorithm(...)
+```
+
+### The Trait
+Below is an abbreviated example of the requestCompression [trait](https://smithy.io/2.0/spec/behavior-traits.html?highlight=requestcompression#requestcompression-trait) in use:
```

Comment:
I still prefer this link to Smithy docs to come earlier in the doc for better reader context. What about in the first sentence? 

`This document covers the design and implementation to support [request compression (https://smithy.io/2.0/spec/behavior-traits.html?highlight=requestcompression#requestcompression-trait).`